Last failure...

> 
> The full traceback is:
> Traceback (most recent call last):
>   File "/tmp/ansible_community.mongodb.mongodb_user_payload_2qjsu79y/ansible_community.mongodb.mongodb_user_payload.zip/ansible_collections/community/mongodb/plugins/modules/mongodb_user.py", line 451, in main
>   File "/tmp/ansible_community.mongodb.mongodb_user_payload_2qjsu79y/ansible_community.mongodb.mongodb_user_payload.zip/ansible_collections/community/mongodb/plugins/modules/mongodb_user.py", line 276, in user_add
>   File "/tmp/ansible_community.mongodb.mongodb_user_payload_2qjsu79y/ansible_community.mongodb.mongodb_user_payload.zip/ansible_collections/community/mongodb/plugins/modules/mongodb_user.py", line 253, in user_find
>   File "/usr/local/lib/python3.10/dist-packages/pymongo/_csot.py", line 125, in csot_wrapper
>     return func(self, *args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/pymongo/synchronous/database.py", line 938, in command
>     return self._command(
>   File "/usr/local/lib/python3.10/dist-packages/pymongo/synchronous/database.py", line 778, in _command
>     return conn.command(
>   File "/usr/local/lib/python3.10/dist-packages/pymongo/synchronous/helpers.py", line 47, in inner
>     return func(*args, **kwargs)
>   File "/usr/local/lib/python3.10/dist-packages/pymongo/synchronous/pool.py", line 398, in command
>     return command(
>   File "/usr/local/lib/python3.10/dist-packages/pymongo/synchronous/network.py", line 212, in command
>     helpers_shared._check_command_response(
>   File "/usr/local/lib/python3.10/dist-packages/pymongo/helpers_shared.py", line 284, in _check_command_response
>     raise OperationFailure(errmsg, code, response, max_wire_version)
> pymongo.errors.OperationFailure: Privilege or restriction details require exact-match usersInfo queries, full error: {'ok': 0.0, 'errmsg': 'Privilege or restriction details require exact-match usersInfo queries', 'code': 20, 'codeName': 'IllegalOperation'}
> fatal: [testhost]: FAILED! => {
>     "changed": false,
>     "invocation": {
>         "module_args": {
>             "atlas_auth": false,
>             "auth_mechanism": null,
>             "authentication_restrictions": [],
>             "connection_options": null,
>             "create_for_localhost_exception": null,
>             "database": "admin",
>             "login_database": "admin",
>             "login_host": "localhost",
>             "login_password": null,
>             "login_port": 3001,
>             "login_user": null,
>             "name": "test_root",
>             "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
>             "replica_set": null,
>             "roles": [
>                 "root"
>             ],
>             "ssl": false,
>             "ssl_ca_certs": null,
>             "ssl_cert_reqs": "CERT_REQUIRED",
>             "ssl_certfile": null,
>             "ssl_crlfile": null,
>             "ssl_keyfile": null,
>             "ssl_pem_passphrase": null,
>             "state": "present",
>             "strict_compatibility": true,
>             "update_password": "always"
>         }
>     },
>     "msg": "Unable to add or update user: Privilege or restriction details require exact-match usersInfo queries, full error: {'ok': 0.0, 'errmsg': 'Privilege or restriction details require exact-match usersInfo queries', 'code': 20, 'codeName': 'IllegalOperation'}"
> }

This is from the mongodb_parameter tests, but from the mongodb_user module...

https://github.com/ansible-collections/community.mongodb/actions/runs/23408451903/job/68095185593?pr=740

--- a/plugins/modules/mongodb_user.py
+++ b/plugins/modules/mongodb_user.py
@@ -245,12 +245,18 @@ def user_find(client, user, db_name):
     Returns:
         dict: when user exists, False otherwise.
     """
-    try:
+    def run_users_info(query_db_name, lookup_db_name):
         users_doc = {
-            'usersInfo': 1,
+            'usersInfo': {
+                'user': user,
+                'db': lookup_db_name,
+            },
             'showAuthenticationRestrictions': True,
         }
-        for mongo_user in client[db_name].command(users_doc)['users']:
+        return client[query_db_name].command(users_doc)['users']
+
+    try:
+        for mongo_user in run_users_info(db_name, db_name):
             if mongo_user['user'] == user:
                 # NOTE: there is no 'db' field in mongo 2.4.
                 if 'db' not in mongo_user:
@@ -258,6 +264,11 @@ def user_find(client, user, db_name):
                 # Workaround to make the condition works with AWS DocumentDB,
                 # since all users are in the admin database.
                 if mongo_user["db"] in [db_name, "admin"]:
+                    return mongo_user
+
+        if db_name != 'admin':
+            for mongo_user in run_users_info('admin', 'admin'):
+                if mongo_user['user'] == user and mongo_user.get('db') == 'admin':
                     return mongo_user
     except Exception as excep:
         if hasattr(excep, 'code') and excep.code == 11:  # 11=UserNotFound

--- a/tests/unit/test_mongodb_user.py
+++ b/tests/unit/test_mongodb_user.py
@@ -139,6 +139,36 @@ class TestMongoDBUserMethods(unittest.TestCase):
         self.assertEqual('alice', user)
         self.assertEqual([], kwargs['authenticationRestrictions'])
 
+    def test_user_find_uses_exact_match_users_info_query(self):
+        class FakeDB:
+            def __init__(self):
+                self.calls = []
+
+            def command(self, payload):
+                self.calls.append(payload)
+                return {
+                    'users': [
+                        {'user': 'alice', 'db': 'admin', 'roles': []},
+                    ]
+                }
+
+        class FakeClient(dict):
+            pass
+
+        db = FakeDB()
+        client = FakeClient({'admin': db})
+
+        result = mongodb_user.user_find(client, 'alice', 'admin')
+
+        self.assertEqual('alice', result['user'])
+        self.assertEqual(
+            {
+                'usersInfo': {'user': 'alice', 'db': 'admin'},
+                'showAuthenticationRestrictions': True,
+            },
+            db.calls[0],
+        )
+
 
 if __name__ == '__main__':
     unittest.main()

